src/v/ssx/checkpoint_mutex.h

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vassert.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/sstring.hh>
+
+#include <utility>
+
+namespace ssx {
+
+template<typename Clock = seastar::timer<>::clock>
+struct basic_mutex_checkpoint {
+    seastar::sstring name;
+    Clock::time_point time;
+};
+
+template<class Clock>
+class basic_mutex_units;
+
+namespace detail {
+struct checkpoint_mutex_locker;
+}
+
+/// Checkpoint mutex allows the user to store some additional information
+/// about the waiter, such as the name of the checkpoint and the time when it
+/// was awaited. In the event of a timeout the exception will contain the
+/// name of the checkpoint that holds the units.
+template<typename Clock = seastar::timer<>::clock>
+class basic_checkpoint_mutex {
```

Comment:
It would be great if this `mutex` had the same api as our current mutex for it be drop in replacement, here the API is completely different.